# DATA INTEGRATION (API DATA AND WEBSCRAPING DATA)

In [ ]:
# import libraries
import pandas as pd
import re
from rapidfuzz import process, fuzz
import unicodedata

# DATA PREPARATION

<br>

WEB SCRAPING DATA (FANTACALCIO.IT)
<br>
<br>

In [ ]:

file_path_scraping = '../1 - DataAcquisition/raw_data/df_players_grades.csv'
df_players_grades_original = pd.read_csv(file_path_scraping)

df_players_grades_original

player_name normalized_player_name team_name        role  grade  \
0            Musso                  Musso  Atalanta  Goalkeeper    6.5   
1       Zappacosta             Zappacosta  Atalanta    Defender    6.5   
2         Djimsiti               Djimsiti  Atalanta    Defender    6.0   
3        Kolasinac              Kolasinac  Atalanta    Defender    6.5   
4           Zortea                 Zortea  Atalanta    Defender    7.0   
...            ...                    ...       ...         ...    ...   
11897   Dani Silva             Dani Silva    Verona  Midfielder    6.0   
11898  Mitrovic S.            S. Mitrovic    Verona  Midfielder    6.5   
11899    Cisse' A.              A. Cisse'    Verona  Midfielder   55.0   
11900       Noslin                 Noslin    Verona    Attacker    7.5   
11901       Tavsan                 Tavsan    Verona    Attacker    6.0   

       fantagrade  goals_scored  goals_conceded  autogoals  penalties_scored  \
0             6.5             0               0          0                 0   
1             6.5             0               0          0                 0   
2             6.0             0               0          0                 0   
3             6.5             0               0          0                 0   
4            10.0             1               0          0                 0   
...           ...           ...             ...        ...               ...   
11897         6.0             0               0          0                 0   
11898         6.5             0               0          0                 0   
11899        55.0             0               0          0                 0   
11900        11.5             1               0          0                 0   
11901         6.0             0               0          0                 0   

       penalties_missed  penalties_saved  assists  player_of_the_match  \
0                     0                0        0                    0   
1                     0                0        0                    0   
2                     0                0        0                    0   
3                     0                0        0                    0   
4                     0                0        0                    0   
...                 ...              ...      ...                  ...   
11897                 0                0        0                    0   
11898                 0                0        0                    0   
11899                 0                0        0                    0   
11900                 0                0        1                    0   
11901                 0                0        0                    0   

        season  matchday  
0      2023/24         1  
1      2023/24         1  
2      2023/24         1  
3      2023/24         1  
4      2023/24         1  
...        ...       ...  
11897  2023/24        38  
11898  2023/24        38  
11899  2023/24        38  
11900  2023/24        38  
11901  2023/24        38  

[11902 rows x 16 columns]

In [ ]:
# order the df by the column 'matchday'
df_players_grades = df_players_grades_original.sort_values(by='matchday')
# perform a groupby on 'player_name' and take the last value of 'team_name' for each group
df_players_grades = df_players_grades.groupby('player_name').agg({'normalized_player_name': 'last','team_name': 'last'}).reset_index()
# add a temporary index
df_players_grades['temporary_player_id'] = df_players_grades.index
# reorder the columns
df_players_grades = df_players_grades[['temporary_player_id', 'player_name', 'normalized_player_name', 'team_name']]

df_players_grades

temporary_player_id player_name normalized_player_name  team_name
0                      0     Abraham                Abraham       Roma
1                      1      Acerbi                 Acerbi      Inter
2                      2        Adli                   Adli      Milan
3                      3       Adopo                  Adopo   Atalanta
4                      4   Aebischer              Aebischer    Bologna
..                   ...         ...                    ...        ...
582                  582   Zielinski              Zielinski     Napoli
583                  583        Zima                   Zima     Torino
584                  584     Zirkzee                Zirkzee    Bologna
585                  585      Zortea                 Zortea  Frosinone
586                  586   Zurkowski              Zurkowski     Empoli

[587 rows x 4 columns]

<br>

API DATA
<br>
<br>

In [ ]:
file_path_api_players = '../1 - DataAcquisition/raw_data/all_players.csv'
file_path_api_teams = '../1 - DataAcquisition/raw_data/serie_a_teams_2024_25.csv'

df_api_players = pd.read_csv(file_path_api_players, index_col=0)
df_api_teams = pd.read_csv(file_path_api_teams, index_col=0)

In [ ]:
# rename the player name and the team name columns
df_api_players = df_api_players.rename(columns={'name': 'player_name'})
df_api_teams = df_api_teams.rename(columns={'name': 'team_name'})


df_api = pd.merge(df_api_players[['player_id', 'player_name', 'team_id']],
                  df_api_teams[['team_id', 'team_name']],
                  how='left', on='team_id').drop(columns='team_id')
df_api

player_id     player_name team_name
0          105  F. Ballo-Touré  AC Milan
1          305        D. Origi  AC Milan
2          765      A. Mirante  AC Milan
3         2045         S. Kjær  AC Milan
4        31054       R. Krunić  AC Milan
..         ...             ...       ...
930      30912      A. Berardi    Verona
931      30922      D. Faraoni    Verona
932      31383      S. Perilli    Verona
933     125715     F. Daniliuc    Verona
934     343570        F. Magro    Verona

[935 rows x 3 columns]

In [ ]:
# team prefixes removal
def clean_team_name(team_name):
    return re.sub(r'^[A-Z]{1,3}\s', '', team_name)

df_api['team_name_cleaned'] = df_api['team_name'].apply(clean_team_name)
df_api['team_name_cleaned'].unique()

array(['Milan', 'Roma', 'Atalanta', 'Bologna', 'Cagliari', 'Como',
       'Empoli', 'Fiorentina', 'Genoa', 'Inter', 'Juventus', 'Lazio',
       'Lecce', 'Monza', 'Napoli', 'Parma', 'Torino', 'Udinese',
       'Venezia', 'Verona'], dtype=object)

<br>

WEB SCRAPING DATA (TRANSFERMARKT.IT)
<br>
<br>

In [ ]:
file_path_api_status_players = '../1 - DataAcquisition/raw_data/df_all_status_players_2024_25.csv'
df_api_status_players = pd.read_csv(file_path_api_status_players)

df_api_status_players

player_id                                  player_status
0          105                              Milan second team
1          305                              Milan second team
2          765  Team of another league or released or retired
3         2045  Team of another league or released or retired
4        31054  Team of another league or released or retired
..         ...                                            ...
930      30912                                         Verona
931      30922                                         Verona
932      31383                                         Verona
933     125715                                         Verona
934     343570                              Verona youth team

[935 rows x 2 columns]

In [ ]:
# condition needed to take only the correct players
conditions = df_api_status_players['player_status'].isin(['Team of another league or released or retired']) | \
             df_api_status_players['player_status'].str.contains('second team|youth team', case=False, na=False)

player_ids = df_api_status_players.loc[conditions, 'player_id'].tolist()

In [ ]:
# filter the API source, keeping only the players who are playing in Serie A this year
df_api = df_api[~df_api['player_id'].isin(player_ids)]
df_api = df_api.drop(columns='team_name')

df_api

player_id   player_name team_name_cleaned
18        1696  S. Chukwueze             Milan
19        1828      L. Jović             Milan
21       19194    T. Abraham             Milan
22       22236   Rafael Leão             Milan
23       22254     Y. Fofana             Milan
..         ...           ...               ...
929      30611    L. Montipò            Verona
930      30912    A. Berardi            Verona
931      30922    D. Faraoni            Verona
932      31383    S. Perilli            Verona
933     125715   F. Daniliuc            Verona

[581 rows x 3 columns]

<br>

NAMES STANDARDIZATION
<br>
<br>

In [ ]:
def add_apostrophe_to_final_accented_vowel(text):
    # set of vowels with grave or acute accents
    accented_vowels = {'à', 'á', 'è', 'é', 'ì', 'í', 'ò', 'ó', 'ù', 'ú'}
    words = text.split()

    # function to add an apostrophe after an accented vowel
    def process_word(word):
        if word[-1] in accented_vowels:
            return f'{word}\''
        return word

    # apply the function to each word
    processed_words = [process_word(word) for word in words]

    return ' '.join(processed_words)

In [ ]:
def diacritic_removal(text):
    # NFD decomposition
    normalized = unicodedata.normalize('NFD', text)

    return ''.join([char for char in normalized if not unicodedata.combining(char)])

In [ ]:
def normalize_special_letters(text):
    # conversion map of special and combined letters
    special_letter_map = {
        'æ': 'ae', 'Æ': 'Ae',
        'œ': 'oe', 'Œ': 'Oe',
        'ß': 'ss',
        'IJ': 'IJ', 'ij': 'ij',
        'Đ': 'D', 'đ': 'd',
        'Ł': 'L', 'ł': 'l',
        'Ø': 'O', 'ø': 'o',
        'Ŋ': 'N', 'ŋ': 'n'
    }

    for letter, replacement in special_letter_map.items():
        text = text.replace(letter, replacement)

    # remove any diacritics with NFD normalization
    normalized = unicodedata.normalize('NFD', text)
    text_without_special_letters = ''.join([char for char in normalized if not unicodedata.combining(char)])

    return text_without_special_letters

In [ ]:
def name_normalization(row):
    # normalization pipeline
    text = row['player_name']
    text = add_apostrophe_to_final_accented_vowel(text)
    text = diacritic_removal(text)
    text = normalize_special_letters(text)
    return text

In [ ]:
df_api['normalized_player_name'] = df_api.apply(name_normalization, axis=1)
df_api = df_api[['player_id', 'player_name', 'normalized_player_name', 'team_name_cleaned']]

df_api

player_id   player_name normalized_player_name team_name_cleaned
18        1696  S. Chukwueze           S. Chukwueze             Milan
19        1828      L. Jović               L. Jovic             Milan
21       19194    T. Abraham             T. Abraham             Milan
22       22236   Rafael Leão            Rafael Leao             Milan
23       22254     Y. Fofana              Y. Fofana             Milan
..         ...           ...                    ...               ...
929      30611    L. Montipò            L. Montipo'            Verona
930      30912    A. Berardi             A. Berardi            Verona
931      30922    D. Faraoni             D. Faraoni            Verona
932      31383    S. Perilli             S. Perilli            Verona
933     125715   F. Daniliuc            F. Daniliuc            Verona

[581 rows x 4 columns]

# DATA INTEGRATION

In [ ]:
df_players_grades.loc[:, 'player_id_matched'] = None
df_api.loc[:, 'player_id_matched'] = None

common_teams = set(df_api['team_name_cleaned']).intersection(set(df_players_grades['team_name']))
common_teams

C:\Users\andre\AppData\Local\Temp\ipykernel_30332\4014844805.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_api.loc[:, 'player_id_matched'] = None


{'Atalanta',
 'Bologna',
 'Cagliari',
 'Empoli',
 'Fiorentina',
 'Genoa',
 'Inter',
 'Juventus',
 'Lazio',
 'Lecce',
 'Milan',
 'Monza',
 'Napoli',
 'Roma',
 'Torino',
 'Udinese',
 'Verona'}

<br>

STEP 1
<br>
<br>

In [ ]:
matching_results_01 = []

# iterate for each common team
for common_team in common_teams:
    # filter the players of that team in the two DataFrames
    players_df_api = df_api[df_api['team_name_cleaned'] == common_team]
    players_df_grades = df_players_grades[df_players_grades['team_name'] == common_team]

    # for each player in the first DataFrame (players_df_api), look for a match in the second DataFrame (players_df_grades)
    for api_idx, row_api in players_df_api.iterrows():
        player_name_api = row_api['normalized_player_name']

        # find the best match with a threshold of 80
        match = process.extractOne(player_name_api, players_df_grades['normalized_player_name'], scorer=fuzz.ratio, score_cutoff=80)

        # if a match exists, update 'player_id_matched'
        if match:
            matched_name, score, matched_idx = match

            matched_player = players_df_grades.loc[matched_idx]

            # add the result to the list for monitoring
            matching_results_01.append({
                'team_name': common_team,
                'player_id_grades': matched_player['temporary_player_id'],
                'player_name_grades': matched_name,
                'player_name_api': player_name_api,
                'player_id_api': row_api['player_id'],
                'matching_score': score
            })

# DataFrame to visualize the matching results
df_matches = pd.DataFrame(matching_results_01)

df_matches

team_name  player_id_grades   player_name_grades      player_name_api  \
0       Inter               163              Dimarco           F. Dimarco   
1       Inter               176             Dumfries          D. Dumfries   
2       Inter               142              Darmian           M. Darmian   
3       Inter               370           Mkhitaryan        H. Mkhitaryan   
4       Inter                86           Calhanoglu        H. Calhanoglu   
..        ...               ...                  ...                  ...   
214    Torino               488              Schuurs           P. Schuurs   
215    Torino               563              Vojvoda           M. Vojvoda   
216    Torino               349               Masina            A. Masina   
217    Torino               298               Lazaro            V. Lazaro   
218    Torino               362  V. Milinkovic-Savic  V. Milinkovic-Savic   

     player_id_api  matching_score  
0            31010       82.352941  
1              226       84.210526  
2              887       82.352941  
3             1457       86.956522  
4             1640       86.956522  
..             ...             ...  
214            535       82.352941  
215           8586       82.352941  
216          18799       80.000000  
217          25353       80.000000  
218          31156      100.000000  

[219 rows x 6 columns]

In [ ]:
print(len(df_matches['player_id_grades']), len(df_matches['player_id_grades'].unique()))

219 219


In [ ]:
print(len(df_matches['player_id_api']), len(df_matches['player_id_api'].unique()))

219 219


In [ ]:
# update df_players_grades and df_api
for _ , row in df_matches.iterrows():
    id_grades = row['player_id_grades']
    id_api = row['player_id_api']

    df_players_grades.loc[df_players_grades['temporary_player_id'] == id_grades, 'player_id_matched'] = id_api
    df_api.loc[df_api['player_id'] == id_api, 'player_id_matched'] = id_grades

<br>

STEP 2
<br>
<br>

In [ ]:
matching_results_02 = []

# iterate for each common team
for common_team in common_teams:
    # filter the players of that team in the two DataFrames (only the players not already matched)
    players_df_api = df_api[(df_api['team_name_cleaned'] == common_team) & (df_api['player_id_matched'].isnull())]
    players_df_grades = df_players_grades[(df_players_grades['team_name'] == common_team) & (df_players_grades['player_id_matched'].isnull())]

    # for each player in the first DataFrame (players_df_api), look for a match in the second DataFrame (players_df_grades)
    for api_idx, row_api in players_df_api.iterrows():
        player_name_api = row_api['normalized_player_name']

        # find the best match with a threshold of 74
        match = process.extractOne(player_name_api, players_df_grades['normalized_player_name'], scorer=fuzz.ratio, score_cutoff=74)

        # if a match exists, update 'player_id_matched'
        if match:
            matched_name, score, matched_idx = match

            matched_player = players_df_grades.loc[matched_idx]

            # add the result to the list for monitoring
            matching_results_02.append({
                'team_name': common_team,
                'player_id_grades': matched_player['temporary_player_id'],
                'player_name_grades': matched_name,
                'player_name_api': player_name_api,
                'player_id_api': row_api['player_id'],
                'matching_score': score
            })

# DataFrame to visualize the matching results
df_matches = pd.DataFrame(matching_results_02)

df_matches

team_name  player_id_grades player_name_grades   player_name_api  \
0      Inter               346        L. Martinez  Lautaro Martinez   
1   Juventus               361              Milik          A. Milik   
2   Juventus               441              Pogba          P. Pogba   
3   Juventus               428              Perin          M. Perin   
4   Juventus               213              Gatti          F. Gatti   
5     Napoli               358              Meret          A. Meret   
6    Udinese               316              Lucca          L. Lucca   
7    Udinese                52              Bijol          J. Bijol   
8    Udinese               398              Okoye          M. Okoye   
9      Milan               263              Jovic          L. Jovic   
10     Milan               377              Musah          Y. Musah   
11     Milan               531              Thiaw          M. Thiaw   
12      Roma               108              Celik          Z. Celik   
13     Monza               339              Maric          M. Maric   
14     Monza                63              Bondo          W. Bondo   
15     Monza               424         P. Pereira     Pedro Pereira   
16  Atalanta               537              Toloi          R. Toloi   
17     Genoa               301              Leali          N. Leali   
18     Lazio               247              Hysaj          E. Hysaj   
19     Lecce               404              Oudin          R. Oudin   
20     Lecce               211              Gallo          A. Gallo   
21     Lecce                30              Banda          L. Banda   
22     Lecce               455              Rafia          H. Rafia   
23     Lecce               172              Dorgu          P. Dorgu   
24    Empoli               232              Gyasi          E. Gyasi   
25    Empoli               330              Maleh          Y. Maleh   
26    Empoli               184              Ekong          E. Ekong   
27   Bologna               444              Posch          S. Posch   
28   Bologna               384              Ndoye          D. Ndoye   
29  Cagliari               576              Zappa          G. Zappa   
30  Cagliari               556              Viola          N. Viola   
31  Cagliari               445              Prati          M. Prati   
32  Cagliari               391              Obert          A. Obert   
33    Torino               484              Savva          Z. Savva   

    player_id_api  matching_score  
0             217       74.074074  
1             333       76.923077  
2             904       76.923077  
3             849       76.923077  
4          268341       76.923077  
5             312       76.923077  
6          199089       76.923077  
7             833       76.923077  
8          143648       76.923077  
9            1828       76.923077  
10         162106       76.923077  
11         163189       76.923077  
12          22222       76.923077  
13           9072       76.923077  
14         266813       76.923077  
15          30768       78.260870  
16          30429       76.923077  
17          31631       76.923077  
18            317       76.923077  
19          22016       76.923077  
20          31543       76.923077  
21         118956       76.923077  
22         136016       76.923077  
23         382452       76.923077  
24          30845       76.923077  
25          56560       76.923077  
26         287374       76.923077  
27            711       76.923077  
28          48648       76.923077  
29            200       76.923077  
30          30629       76.923077  
31         309388       76.923077  
32         321744       76.923077  
33         343332       76.923077

In [ ]:
print(len(df_matches['player_id_grades']), len(df_matches['player_id_grades'].unique()))

34 34


In [ ]:
# update df_players_grades and df_api
for _ , row in df_matches.iterrows():
    id_grades = row['player_id_grades']
    id_api = row['player_id_api']

    df_players_grades.loc[df_players_grades['temporary_player_id'] == id_grades, 'player_id_matched'] = id_api
    df_api.loc[df_api['player_id'] == id_api, 'player_id_matched'] = id_grades

In [ ]:
df_api[df_api['normalized_player_name']=='Josep Martinez']

player_id     player_name normalized_player_name team_name_cleaned  \
483      46988  Josep Martínez         Josep Martinez             Inter   

    player_id_matched  
483              None

In [ ]:
df_api[df_api['normalized_player_name']=='Lautaro Martinez']

player_id       player_name normalized_player_name team_name_cleaned  \
456        217  Lautaro Martínez       Lautaro Martinez             Inter   

    player_id_matched  
456               346

In [ ]:
df_players_grades[df_players_grades['normalized_player_name']=='L. Martinez']

temporary_player_id  player_name normalized_player_name team_name  \
346                  346  Martinez L.            L. Martinez     Inter   

    player_id_matched  
346               217

<br>

STEP 3
<br>
<br>

In [ ]:
matching_results_03 = []

# iterate for each common team
for common_team in common_teams:
    # filter the players of that team in the two DataFrames (only the players not already matched)
    players_df_api = df_api[(df_api['team_name_cleaned'] == common_team) & (df_api['player_id_matched'].isnull())]
    players_df_grades = df_players_grades[(df_players_grades['team_name'] == common_team) & (df_players_grades['player_id_matched'].isnull())]

    # for each player in the first DataFrame (players_df_api), look for a match in the second DataFrame (players_df_grades)
    for api_idx, row_api in players_df_api.iterrows():
        player_name_api = row_api['normalized_player_name']

        # find the best match with a threshold of 70
        match = process.extractOne(player_name_api, players_df_grades['normalized_player_name'], scorer=fuzz.ratio, score_cutoff=70)

        # if a match exists, update 'player_id_matched'
        if match:
            matched_name, score, matched_idx = match

            matched_player = players_df_grades.loc[matched_idx]

            # add the result to the list for monitoring
            matching_results_03.append({
                'team_name': common_team,
                'player_id_grades': matched_player['temporary_player_id'],
                'player_name_grades': matched_name,
                'player_name_api': player_name_api,
                'player_id_api': row_api['player_id'],
                'matching_score': score
            })

# DataFrame to visualize the matching results
df_matches = pd.DataFrame(matching_results_03)

df_matches

team_name  player_id_grades player_name_grades player_name_api  \
0      Inter               149            De Vrij      S. de Vrij   
1   Juventus               356           Mckennie     W. McKennie   
2   Juventus               567               Weah         T. Weah   
3     Napoli               573     Zambo Anguissa     F. Anguissa   
4    Udinese               578            Zarraga    Oier Zarraga   
5      Monza               167             Djuric        M. Duric   
6      Monza               257               Izzo         A. Izzo   
7   Atalanta               182       D.s. Ederson         Ederson   
8   Atalanta               146            De Roon      M. de Roon   
9   Atalanta               243               Hien         I. Hien   
10     Genoa                31               Bani         M. Bani   
11     Lecce               265               Kaba         M. Kaba   
12    Empoli               234               Haas         N. Haas   
13    Verona               175               Duda         O. Duda   
14    Verona               135               Cruz         J. Cruz   
15  Cagliari               322            Luvumbo    Zito Luvumbo   
16  Cagliari               363               Mina         Y. Mina   
17    Torino               251               Ilic         I. Ilic   

    player_id_api  matching_score  
0             194       70.588235  
1             415       73.684211  
2            1138       72.727273  
3            3406       72.000000  
4          182560       73.684211  
5           31692       71.428571  
6           30501       72.727273  
7           10097       73.684211  
8           30432       70.588235  
9          137976       72.727273  
10          30736       72.727273  
11         197779       72.727273  
12          31555       72.727273  
13          25349       72.727273  
14         129892       72.727273  
15         140831       73.684211  
16           2484       72.727273  
17          46170       72.727273

In [ ]:
print(len(df_matches['player_id_api']), len(df_matches['player_id_api'].unique()))

18 18


In [ ]:
# update df_players_grades and df_api
for _ , row in df_matches.iterrows():
    id_grades = row['player_id_grades']
    id_api = row['player_id_api']

    df_players_grades.loc[df_players_grades['temporary_player_id'] == id_grades, 'player_id_matched'] = id_api
    df_api.loc[df_api['player_id'] == id_api, 'player_id_matched'] = id_grades

<br>

STEP 4
<br>
<br>

In [ ]:
matching_results_04 = []

# iterate for each common team
for common_team in common_teams:
    # filter the players of that team in the two DataFrames (only the players not already matched)
    players_df_api = df_api[(df_api['team_name_cleaned'] == common_team) & (df_api['player_id_matched'].isnull())]
    players_df_grades = df_players_grades[(df_players_grades['team_name'] == common_team) & (df_players_grades['player_id_matched'].isnull())]

    # for each player in the first DataFrame (players_df_api), look for a match in the second DataFrame (players_df_grades)
    for api_idx, row_api in players_df_api.iterrows():
        player_name_api = row_api['normalized_player_name']

        # find the best match with a threshold of 50
        match = process.extractOne(player_name_api, players_df_grades['normalized_player_name'], scorer=fuzz.ratio, score_cutoff=50)

        # if a match exists, update 'player_id_matched'
        if match:
            matched_name, score, matched_idx = match

            matched_player = players_df_grades.loc[matched_idx]

            # add the result to the list for monitoring
            matching_results_04.append({
                'team_name': common_team,
                'player_id_grades': matched_player['temporary_player_id'],
                'player_name_grades': matched_name,
                'player_name_api': player_name_api,
                'player_id_api': row_api['player_id'],
                'matching_score': score
            })

# DataFrame to visualize the matching results
df_matches = pd.DataFrame(matching_results_04)

df_matches

team_name  player_id_grades player_name_grades  player_name_api  \
0     Juventus               388              Nonge      N. Gonzalez   
1     Juventus               570             Yildiz        K. Yıldız   
2      Udinese                 5            M. Ake'        I. Toure'   
3        Milan               469          L. Romero      L. Torriani   
4         Roma               308        D. Llorente        E. Le Fee   
5         Roma               308        D. Llorente         M. Kone'   
6         Roma               380            N'Dicka        E. Ndicka   
7        Monza               374               Mota        Dany Mota   
8        Monza               508      A. Sorrentino       A. Petagna   
9        Monza                97         V. Carboni        O. Forson   
10       Monza               508      A. Sorrentino        A. Bianco   
11       Monza                97         V. Carboni        M. Valoti   
12       Monza               508      A. Sorrentino         S. Sensi   
13       Monza               338              Mari'      Pablo Mari'   
14       Monza               508      A. Sorrentino        A. Cragno   
15    Atalanta               540          E. Toure'       B. Godfrey   
16       Genoa               359            Messias   Junior Messias   
17       Genoa               345       Jo. Martinez       F. Miretti   
18       Genoa               345       Jo. Martinez        P. Masini   
19       Genoa               343             Martin     Aaron Martin   
20       Lazio               196    Felipe Anderson  Andre' Anderson   
21       Lazio               219               Gila       Mario Gila   
22       Lazio               196    Felipe Anderson    Filipe Bordon   
23       Lecce               157    F. Di Francesco    F. Brancolini   
24  Fiorentina               313           M. Lopez          M. Kean   
25  Fiorentina               227        N. Gonzalez        R. Gosens   
26  Fiorentina               313           M. Lopez          E. Bove   
27  Fiorentina               313           M. Lopez        M. Moreno   
28      Empoli                39         S. Bastoni      S. Esposito   
29      Empoli               129          G. Corona       L. Colombo   
30      Empoli                90        Cancellieri      P. Pellegri   
31      Empoli               129          G. Corona       T. Anjorin   
32      Empoli                39         S. Bastoni       S. Sazonov   
33      Empoli               458       F. Ranocchia    L. Marianucci   
34      Empoli                88          Cambiaghi        J. Sambia   
35      Empoli               496         S. Shpendi      J. Seghetti   
36      Verona               109           Centonze     C. Tengstedt   
37      Verona               389             Noslin         Joselito   
38    Cagliari                23               Azzi       Paulo Azzi   
39      Torino               420           Pellegri       A. Paleari   
40      Torino               485            Sazonov        Saul Coco   

    player_id_api  matching_score  
0           26315       50.000000  
1          339883       53.333333  
2          195893       50.000000  
3          386298       50.000000  
4           20638       50.000000  
5           22147       52.631579  
6            1807       62.500000  
7           30603       61.538462  
8           30879       52.173913  
9          284242       52.631579  
10         302432       54.545455  
11          30875       52.631579  
12          30535       57.142857  
13          46792       62.500000  
14          30546       54.545455  
15          19073       52.631579  
16          56396       66.666667  
17         181808       54.545455  
18         281096       57.142857  
19          25914       66.666667  
20          31688       66.666667  
21         162952       57.142857  
22         407916       64.285714  
23          30391       57.142857  
24            877       53.333333  
25          30422       50.000000  
26      

OK:

Lazio	219	Gila	Mario Gila	162952

Genoa	359	Messias	Junior Messias	56396

Juventus	570	Yildiz	K. Yıldız	339883

Monza	374	Mota	Dany Mota	30603

Monza	338	Mari'	Pablo Mari'	46792

Cagliari	23	Azzi	Paulo Azzi	56207

Roma	380	  N'Dicka 	E. Ndicka	1807

In [ ]:
filtered_df_matches = df_matches[df_matches['player_id_grades'].isin([219, 359, 570, 374, 338, 23, 380])]

In [ ]:
print(len(filtered_df_matches['player_id_grades']), len(filtered_df_matches['player_id_grades'].unique()))

7 7


In [ ]:
# update df_players_grades and df_api
for _ , row in filtered_df_matches.iterrows():
    id_grades = row['player_id_grades']
    id_api = row['player_id_api']

    df_players_grades.loc[df_players_grades['temporary_player_id'] == id_grades, 'player_id_matched'] = id_api
    df_api.loc[df_api['player_id'] == id_api, 'player_id_matched'] = id_grades

<br>

STEP 5
<br>
<br>

In [ ]:
matching_results_05 = []


# filter the players who have not yet been matched
players_df_api = df_api[df_api['player_id_matched'].isnull()]
players_df_grades = df_players_grades[df_players_grades['player_id_matched'].isnull()]

# for each player in the first DataFrame (players_df_api), look for a match in the second DataFrame (players_df_grades)
for api_idx, row_api in players_df_api.iterrows():
    player_name_api = row_api['normalized_player_name']

    # find the best match in the first DataFrame with a threshold of 80
    match = process.extractOne(player_name_api, players_df_grades['normalized_player_name'], scorer=fuzz.ratio, score_cutoff=80)

    # if a match exists, update 'player_id_matched'
    if match:
        matched_name, score, matched_idx = match

        matched_player = players_df_grades.loc[matched_idx]

        # add the result to the list for monitoring
        matching_results_05.append({
            'team_name_grades': matched_player['team_name'],
            'player_id_grades': matched_player['temporary_player_id'],
            'player_name_grades': matched_name,
            'player_name_api': player_name_api,
            'player_id_api': row_api['player_id'],
            'team_name_api': row_api['team_name_cleaned'],
            'matching_score': score
        })

# DataFrame to visualize the matching results
df_matches = pd.DataFrame(matching_results_05)

df_matches

team_name_grades  player_id_grades player_name_grades  player_name_api  \
0              Roma                 0            Abraham       T. Abraham   
1           Bologna               477       Saelemaekers  A. Saelemaekers   
2          Cagliari               495         Shomurodov    E. Shomurodov   
3         Frosinone               510             Soule'        M. Soule'   
4         Frosinone                74        Brescianini   M. Brescianini   
..              ...               ...                ...              ...   
72        Frosinone               236            Harroui       A. Harroui   
73      Salernitana                70           Bradaric      D. Bradaric   
74         Sassuolo                45            Berardi       A. Berardi   
75       Fiorentina               193            Faraoni       D. Faraoni   
76      Salernitana               140           Daniliuc      F. Daniliuc   

    player_id_api team_name_api  matching_score  
0           19194         Milan       82.352941  
1            1417          Roma       88.888889  
2           53535          Roma       86.956522  
3          323936          Roma       80.000000  
4            1639      Atalanta       88.000000  
..            ...           ...             ...  
72          37437        Verona       82.352941  
73          14327        Verona       84.210526  
74          30912        Verona       82.352941  
75          30922        Verona       82.352941  
76         125715        Verona       84.210526  

[77 rows x 7 columns]


NO:

Atalanta	175	E. Toure'	I. Toure'	195893	Udinese

Juventus	299	L. Cerri	A. Cerri	30570	Como

Sassuolo	45	Berardi	A. Berardi	30912	Verona

In [ ]:
filtered_df_matches = df_matches[~df_matches['player_id_grades'].isin([175, 299, 45])]

In [ ]:
print(len(filtered_df_matches['player_id_grades']), len(filtered_df_matches['player_id_grades'].unique()))

76 74


In [ ]:
duplicates = filtered_df_matches['player_id_grades'][filtered_df_matches['player_id_grades'].duplicated()]

duplicate_rows = filtered_df_matches[filtered_df_matches['player_id_grades'].isin(duplicates)]

duplicate_rows

team_name_grades  player_id_grades player_name_grades player_name_api  \
40       Fiorentina               227        N. Gonzalez     N. Gonzalez   
47       Fiorentina               227        N. Gonzalez     D. Gonzalez   
49      Salernitana               131       L. Coulibaly    L. Coulibaly   
63      Salernitana               131       L. Coulibaly    W. Coulibaly   

    player_id_api team_name_api  matching_score  
40          26315      Juventus      100.000000  
47         307103         Lazio       90.909091  
49           1748         Lecce      100.000000  
63         128338         Parma       91.666667

In [ ]:
filtered_df_matches = filtered_df_matches[~filtered_df_matches['player_id_api'].isin([307103, 128338])]

In [ ]:
print(len(filtered_df_matches['player_id_grades']), len(filtered_df_matches['player_id_grades'].unique()))

74 74


In [ ]:
print(len(filtered_df_matches['player_id_api']), len(filtered_df_matches['player_id_api'].unique()))

74 74


In [ ]:
# update df_players_grades and df_api
for _ , row in filtered_df_matches.iterrows():
    id_grades = row['player_id_grades']
    id_api = row['player_id_api']

    df_players_grades.loc[df_players_grades['temporary_player_id'] == id_grades, 'player_id_matched'] = id_api
    df_api.loc[df_api['player_id'] == id_api, 'player_id_matched'] = id_grades

<br>

STEP 6
<br>
<br>

In [ ]:
matching_results_06 = []


# filter the players who have not yet been matched
players_df_api = df_api[(df_api['player_id_matched'].isnull())]
players_df_grades = df_players_grades[df_players_grades['player_id_matched'].isnull()]

# for each player in the first DataFrame (players_df_api), look for a match in the second DataFrame (players_df_grades)
for api_idx, row_api in players_df_api.iterrows():
    player_name_api = row_api['normalized_player_name']

    # find the best match in the first DataFrame with a threshold of 72
    match = process.extractOne(player_name_api, players_df_grades['normalized_player_name'], scorer=fuzz.ratio, score_cutoff=72)

    # if a match exists, update 'player_id_matched'
    if match:
        matched_name, score, matched_idx = match

        matched_player = players_df_grades.loc[matched_idx]

        # add the result to the list for monitoring
        matching_results_06.append({
            'team_name_grades': matched_player['team_name'],
            'player_id_grades': matched_player['temporary_player_id'],
            'player_name_grades': matched_name,
            'player_name_api': player_name_api,
            'player_id_api': row_api['player_id'],
            'team_name_api': row_api['team_name_cleaned'],
            'matching_score': score
        })
# DataFrame to visualize the matching results
df_matches = pd.DataFrame(matching_results_06)
df_matches = df_matches[df_matches['matching_score']<80]

df_matches

team_name_grades  player_id_grades player_name_grades player_name_api  \
0            Torino               507              Soppy        B. Soppy   
1          Sassuolo               365         K. Miranda    Juan Miranda   
2          Sassuolo               189              Erlic        M. Erlic   
3          Atalanta               244               Holm         E. Holm   
4            Empoli               340              Marin        R. Marin   
5          Atalanta                 3              Adopo        M. Adopo   
6            Torino               551              Verdi        S. Verdi   
7          Sassuolo               557               Viti         M. Viti   
8          Juventus               277               Kean         M. Kean   
9             Milan                 2               Adli         Y. Adli   
10             Roma                69               Bove         E. Bove   
11           Verona                79              Cabal        J. Cabal   
12      Salernitana                37              Basic        T. Basic   
13           Empoli               496         S. Shpendi        S. Sensi   
14            Milan                87            Camarda       D. Camara   
16            Genoa               235               Haps         R. Haps   

    player_id_api team_name_api  matching_score  
0          162347      Atalanta       76.923077  
1             134       Bologna       72.727273  
2           30827       Bologna       76.923077  
3           47985       Bologna       72.727273  
4            2118      Cagliari       76.923077  
5           30505      Cagliari       76.923077  
6             334          Como       76.923077  
7          180510        Empoli       72.727273  
8             877    Fiorentina       72.727273  
9            1265    Fiorentina       72.727273  
10         286475    Fiorentina       72.727273  
11         125674      Juventus       76.923077  
12           1266         Lazio       76.923077  
13          30535         Monza       77.777778  
14          91424         Parma       75.000000  
16          37144       Venezia       72.727273

NO:

Sassuolo	365	K. Miranda	Juan Miranda	134	Bologna

Empoli	496	S. Shpendi	S. Sensi	30535	Monza

Milan	87	Camarda	D. Camara	91424	Parma

In [ ]:
filtered_df_matches = df_matches[~df_matches['player_id_grades'].isin([365, 496, 87])]

In [ ]:
print(len(filtered_df_matches['player_id_grades']), len(filtered_df_matches['player_id_grades'].unique()))

13 13


In [ ]:
# update df_players_grades and df_api
for _ , row in filtered_df_matches.iterrows():
    id_grades = row['player_id_grades']
    id_api = row['player_id_api']

    df_players_grades.loc[df_players_grades['temporary_player_id'] == id_grades, 'player_id_matched'] = id_api
    df_api.loc[df_api['player_id'] == id_api, 'player_id_matched'] = id_grades

<br>

STEP 7
<br>
<br>

In [ ]:
matching_results_07 = []


# filter the players who have not yet been matched
players_df_api = df_api[(df_api['player_id_matched'].isnull()) & (~df_api['player_id'].isin([134, 30535, 91424]))]
players_df_grades = df_players_grades[df_players_grades['player_id_matched'].isnull()]

# for each player in the first DataFrame (players_df_api), look for a match in the second DataFrame (players_df_grades)
for api_idx, row_api in players_df_api.iterrows():
    player_name_api = row_api['normalized_player_name']

    # find the best match in the first DataFrame with a threshold of 60
    match = process.extractOne(player_name_api, players_df_grades['normalized_player_name'], scorer=fuzz.ratio, score_cutoff=60)

    # if a match exists, update 'player_id_matched'
    if match:
        matched_name, score, matched_idx = match

        matched_player = players_df_grades.loc[matched_idx]

        # add the result to the list for monitoring
        matching_results_07.append({
            'team_name_grades': matched_player['team_name'],
            'player_id_grades': matched_player['temporary_player_id'],
            'player_name_grades': matched_name,
            'player_name_api': player_name_api,
            'player_id_api': row_api['player_id'],
            'team_name_api': row_api['team_name_cleaned'],
            'matching_score': score
        })

# DataFrame to visualize the matching results
df_matches = pd.DataFrame(matching_results_07)
df_matches = df_matches[df_matches['matching_score'] <= 72]

OK:

Lecce	515	Strefezza	Gabriel Strefezza	31624	Como

Genoa	343	Martin	Aaron Martin	25914	Genoa

Salernitana	162	Dia	B. Dia	22015.0	Lazio

Fiorentina	18	Arthur Melo	  Arthur	150	Juventus

Juventus	387	Nicolussi Caviglia	H. Nicolussi	881	Venezia

In [ ]:
filtered_df_matches = df_matches[(df_matches['player_id_grades'].isin([515, 343, 162, 18, 387]))]

In [ ]:
print(len(filtered_df_matches['player_id_grades']), len(filtered_df_matches['player_id_grades'].unique()))

7 5


In [ ]:
filtered_df_matches

team_name_grades  player_id_grades  player_name_grades    player_name_api  \
8             Lecce               515           Strefezza  Gabriel Strefezza   
17            Genoa               343              Martin       Aaron Martin   
20       Fiorentina                18         Arthur Melo             Arthur   
23      Salernitana               162                 Dia             B. Dia   
33      Salernitana               162                 Dia            D. Diaw   
34            Genoa               343              Martin         Rafa Marin   
46         Juventus               387  Nicolussi Caviglia       H. Nicolussi   

    player_id_api team_name_api  matching_score  
8           31624          Como       69.230769  
17          25914         Genoa       66.666667  
20            150      Juventus       70.588235  
23          22015         Lazio       66.666667  
33          30663         Monza       60.000000  
34         237066        Napoli       62.500000  
46            881       Venezia       60.000000

In [ ]:
filtered_df_matches = filtered_df_matches[~filtered_df_matches['player_id_api'].isin([30663, 237066])]

In [ ]:
print(len(filtered_df_matches['player_id_grades']), len(filtered_df_matches['player_id_grades'].unique()))

5 5


In [ ]:
print(len(filtered_df_matches['player_id_api']), len(filtered_df_matches['player_id_api'].unique()))

5 5


In [ ]:
# update df_players_grades and df_api
for _ , row in filtered_df_matches.iterrows():
    id_grades = row['player_id_grades']
    id_api = row['player_id_api']

    df_players_grades.loc[df_players_grades['temporary_player_id'] == id_grades, 'player_id_matched'] = id_api
    df_api.loc[df_api['player_id'] == id_api, 'player_id_matched'] = id_grades

<br>

STEP 8
<br>
<br>

In [ ]:
matching_results_08 = []


# filter the players who have not yet been matched
players_df_api = df_api[(df_api['player_id_matched'].isnull()) & (~df_api['player_id'].isin([134, 30535, 91424]))]
players_df_grades = df_players_grades[df_players_grades['player_id_matched'].isnull()]

# for each player in the first DataFrame (players_df_api), look for a match in the second DataFrame (players_df_grades)
for api_idx, row_api in players_df_api.iterrows():
    player_name_api = row_api['normalized_player_name']

    # find the best match in the first DataFrame with a threshold of 60
    match = process.extractOne(player_name_api, players_df_grades['normalized_player_name'], scorer=fuzz.ratio, score_cutoff=50)

    # if a match exists, update 'player_id_matched'
    if match:
        matched_name, score, matched_idx = match

        matched_player = players_df_grades.loc[matched_idx]

        # add the result to the list for monitoring
        matching_results_08.append({
            'team_name_grades': matched_player['team_name'],
            'player_id_grades': matched_player['temporary_player_id'],
            'player_name_grades': matched_name,
            'player_name_api': player_name_api,
            'player_id_api': row_api['player_id'],
            'team_name_api': row_api['team_name_cleaned'],
            'matching_score': score
        })

# DataFrame to visualize the matching results
df_matches = pd.DataFrame(matching_results_08)
df_matches = df_matches[df_matches['matching_score'] <= 60]

<br>

NOT MATCHED PLAYERS
<br>
<br>

In [ ]:
len(df_players_grades[df_players_grades['player_id_matched'].isnull()])

217

In [ ]:
len(df_api[df_api['player_id_matched'].isnull()])

211

# NAME HARMONIZATION

In [ ]:
df_players_grades_original['api_player_id'] = None

# filter the df (only matched players)
df_players_grades_not_nulls = df_players_grades[~df_players_grades['player_id_matched'].isnull()]

for idx, row in df_players_grades_not_nulls.iterrows():
    # player name and team name (from df_players_grades source) extraction
    player_name = row['player_name']
    team_name = row['team_name']

    # player name (from df_api source) extraction
    player_id_matched = row['player_id_matched']
    condition = df_api['player_id']==player_id_matched
    api_player_name = df_api.loc[condition, 'player_name'].iloc[0]

    # substitution of the name in df_players_grades_original
    conditions = (df_players_grades_original['player_name'] == player_name)

    df_players_grades_original.loc[conditions, 'player_name'] = api_player_name
    df_players_grades_original.loc[conditions, 'api_player_id'] = player_id_matched

df_players_grades_original = df_players_grades_original[['api_player_id', 'player_name',
                                                         'normalized_player_name', 'team_name', 'role',
                                                         'grade', 'fantagrade', 'goals_scored',
                                                         'goals_conceded', 'autogoals', 'penalties_scored',
                                                         'penalties_missed', 'penalties_saved', 'assists',
                                                         'player_of_the_match', 'season', 'matchday']]

df_players_grades_original

api_player_id    player_name normalized_player_name team_name  \
0              None          Musso                  Musso  Atalanta   
1              2286  D. Zappacosta             Zappacosta  Atalanta   
2             30421    B. Djimsiti               Djimsiti  Atalanta   
3              1442   S. Kolašinac              Kolasinac  Atalanta   
4            128461      N. Zortea                 Zortea  Atalanta   
...             ...            ...                    ...       ...   
11897         42007     Dani Silva             Dani Silva    Verona   
11898          None    Mitrovic S.            S. Mitrovic    Verona   
11899        359100       A. Cissè              A. Cisse'    Verona   
11900        133729      T. Noslin                 Noslin    Verona   
11901          None         Tavsan                 Tavsan    Verona   

             role  grade  fantagrade  goals_scored  goals_conceded  autogoals  \
0      Goalkeeper    6.5         6.5             0               0          0   
1        Defender    6.5         6.5             0               0          0   
2        Defender    6.0         6.0             0               0          0   
3        Defender    6.5         6.5             0               0          0   
4        Defender    7.0        10.0             1               0          0   
...           ...    ...         ...           ...             ...        ...   
11897  Midfielder    6.0         6.0             0               0          0   
11898  Midfielder    6.5         6.5             0               0          0   
11899  Midfielder   55.0        55.0             0               0          0   
11900    Attacker    7.5        11.5             1               0          0   
11901    Attacker    6.0         6.0             0               0          0   

       penalties_scored  penalties_missed  penalties_saved  assists  \
0                     0                 0                0        0   
1                     0                 0                0        0   
2                     0                 0                0        0   
3                     0                 0                0        0   
4                     0                 0                0        0   
...                 ...               ...              ...      ...   
11897                 0                 0                0        0   
11898                 0                 0                0        0   
11899                 0                 0                0        0   
11900                 0                 0                0        1   
11901                 0                 0                0        0   

       player_of_the_match   season  matchday  
0                        0  2023/24         1  
1                        0  2023/24         1  
2                        0  2023/24         1  
3                        0  2023/24         1  
4                        0  2023/24         1  
...                    ...      ...       ...  
11897                    0  2023/24        38  
11898                    0  2023/24        38  
11899                    0  2023/24        38  
11900                    0  2023/24        38  
11901                    0  2023/24        38  

[11902 rows x 17 columns]

In [ ]:
len(set(df_players_grades_original['api_player_id']) - set([None]))

370

In [ ]:
df_players_grades_original.to_csv('../1 - DataAcquisition/raw_data/df_players_grades_post_integration.csv',
                                  index=False)

In [ ]:
df_players_grades_original

api_player_id    player_name normalized_player_name team_name  \
0              None          Musso                  Musso  Atalanta   
1              2286  D. Zappacosta             Zappacosta  Atalanta   
2             30421    B. Djimsiti               Djimsiti  Atalanta   
3              1442   S. Kolašinac              Kolasinac  Atalanta   
4            128461      N. Zortea                 Zortea  Atalanta   
...             ...            ...                    ...       ...   
11897         42007     Dani Silva             Dani Silva    Verona   
11898          None    Mitrovic S.            S. Mitrovic    Verona   
11899        359100       A. Cissè              A. Cisse'    Verona   
11900        133729      T. Noslin                 Noslin    Verona   
11901          None         Tavsan                 Tavsan    Verona   

             role  grade  fantagrade  goals_scored  goals_conceded  autogoals  \
0      Goalkeeper    6.5         6.5             0               0          0   
1        Defender    6.5         6.5             0               0          0   
2        Defender    6.0         6.0             0               0          0   
3        Defender    6.5         6.5             0               0          0   
4        Defender    7.0        10.0             1               0          0   
...           ...    ...         ...           ...             ...        ...   
11897  Midfielder    6.0         6.0             0               0          0   
11898  Midfielder    6.5         6.5             0               0          0   
11899  Midfielder   55.0        55.0             0               0          0   
11900    Attacker    7.5        11.5             1               0          0   
11901    Attacker    6.0         6.0             0               0          0   

       penalties_scored  penalties_missed  penalties_saved  assists  \
0                     0                 0                0        0   
1                     0                 0                0        0   
2                     0                 0                0        0   
3                     0                 0                0        0   
4                     0                 0                0        0   
...                 ...               ...              ...      ...   
11897                 0                 0                0        0   
11898                 0                 0                0        0   
11899                 0                 0                0        0   
11900                 0                 0                0        1   
11901                 0                 0                0        0   

       player_of_the_match   season  matchday  
0                        0  2023/24         1  
1                        0  2023/24         1  
2                        0  2023/24         1  
3                        0  2023/24         1  
4                        0  2023/24         1  
...                    ...      ...       ...  
11897                    0  2023/24        38  
11898                    0  2023/24        38  
11899                    0  2023/24        38  
11900                    0  2023/24        38  
11901                    0  2023/24        38  

[11902 rows x 17 columns]